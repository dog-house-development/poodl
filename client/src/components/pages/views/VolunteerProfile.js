import React, { Component } from 'react';
import { connect } from 'react-redux';
import _ from 'lodash';
import { fetchVolunteer } from '../../../actions/volunteerActions';
import Loading from '../../ui/Loading';
import { Link } from 'react-router-dom';

export class VolunteerProfile extends Component {
    constructor(props) {
        super(props);
        this.routeParam = props.match.params.id;
    }

    componentDidMount() {
        // call redux action to retrieve specified volunteer from api
        this.props.getVolunteer(this.routeParam);
    }

    getPageMarkup() {
        if (this.props.loading) {
            return <Loading content="Loading volunteer info..." />;
        } else {
            return (
                <h1>
                    {_.get(this.props.volunteer, 'firstName')} {_.get(this.props.volunteer, 'lastName')}
                </h1>
            );
        }
    }

    render() {
        return (
            <div className="view-all-container">
                <Link to="/volunteers" className="button small tertiary">
                    <i className="material-icons">keyboard_backspace</i> Back to all volunteers
                </Link>
                <div className="panel">{this.getPageMarkup()}</div>
            </div>
        );
    }
}

export const mapStateToProps = (state, props) => {
    return {
        volunteer: state.volunteers.one,
        errors: state.errors
    };
};

export const mapDispatchToProps = dispatch => {
    return {
        getVolunteer: id => dispatch(fetchVolunteer(id))
    };
};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(VolunteerProfile);
