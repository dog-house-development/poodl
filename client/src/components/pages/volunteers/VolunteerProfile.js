import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import _ from 'lodash';
import VolunteerActions from '../../../actions/volunteerActions';
import { Link } from 'react-router-dom';
import DynamicForm from '../../ui/DynamicForm';
import volunteerInputs from './volunteerInputs';
import { DeleteButton } from './../../ui/DeleteButton';
import Loading from './../../ui/Loading';

export class VolunteerProfile extends Component {
    static defaultProps = {
        errors: {}
    };

    componentDidMount() {
        // call redux action to retrieve specified profile from api
        this.props.volunteerActions.get(this.props.match.params.id, () => {
            this.props.history.push('/pageNotFound');
        });
        window.scrollTo(0, 0);
    }

    editVolunteer = (modifiedInputs, onSuccess) => {
        this.props.volunteerActions.edit(_.get(this.props.volunteer, '_id'), modifiedInputs, onSuccess);
    };

    getVolunteerName() {
        return this.props.loading ? (
            <Loading content="" />
        ) : (
            _.get(this.props.volunteer, 'firstName') + ' ' + _.get(this.props.volunteer, 'lastName')
        );
    }

    handleDeleteClick = () => {
        this.props.volunteerActions.delete(this.props.match.params.id, () => this.props.history.push('/volunteers'));
    };

    render() {
        return (
            <div className="page-container">
                <Link to="/volunteers" className="button small tertiary icon">
                    <i className="material-icons button-icon">keyboard_backspace</i> Back to all volunteers
                </Link>
                <div>
                    <div className="page-header">
                        <h1>{this.getVolunteerName()}</h1>
                        <DeleteButton
                            onConfirm={this.handleDeleteClick}
                            confirmQuestion={`Are you sure you want to delete the volunteer '${this.getVolunteerName()}'?`}>
                            Delete Volunteer
                        </DeleteButton>
                    </div>

                    <DynamicForm
                        inputs={volunteerInputs}
                        editValues={this.editVolunteer}
                        values={this.props.volunteer}
                        editable={true}
                        loading={this.props.loading}
                        errors={this.props.errors}
                    />
                </div>
            </div>
        );
    }
}

export const mapStateToProps = (state, props) => {
    return {
        volunteer: state.volunteers.all[props.match.params.id],
        loading: state.volunteers.loading,
        errors: state.volunteers.errors
    };
};

export const mapDispatchToProps = dispatch => {
    return {
        volunteerActions: bindActionCreators(VolunteerActions, dispatch)
    };
};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(VolunteerProfile);
