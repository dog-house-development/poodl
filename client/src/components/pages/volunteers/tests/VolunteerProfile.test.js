import React from 'react';
import ReactDOM from 'react-dom';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import _ from 'lodash';
import { Link } from 'react-router-dom';
import { VolunteerProfile, mapStateToProps, mapDispatchToProps } from '../VolunteerProfile';

configure({ adapter: new Adapter() });

const match = {
    params: {
        _id: '123' //any id you want to set
    }
};
describe('VolunteerProfile tests', () => {
    global.scrollTo = jest.fn();
    let state, props, wrapper, instance;
    const setInstanceAndWrapper = (_props = {}, _state = {}) => {
        state = _.assign(
            {},
            {
                volunteers: {
                    all: {
                        123: {
                            _id: '123',
                            firstName: 'Big',
                            lastName: 'Tup',
                            email: 'bigtup@nowhere.com'
                        }
                    },
                    errors: {}
                }
            },
            _state
        );
        props = _.assign({}, { match: { params: { id: 123 } } }, _props);
        wrapper = shallow(
            <VolunteerProfile
                match={match}
                {..._.assign(
                    {},
                    props,
                    mapStateToProps(state, props),
                    mapDispatchToProps(jasmine.createSpy('dispatch'))
                )}
            />
        );
        instance = wrapper.instance();
    };

    beforeEach(() => {
        setInstanceAndWrapper();
    });

    describe('mapStateToProps', () => {
        it('should map state to props', () => {
            expect(mapStateToProps(state, props)).toEqual({
                volunteer: {
                    _id: '123',
                    firstName: 'Big',
                    lastName: 'Tup',
                    email: 'bigtup@nowhere.com'
                },
                errors: {}
            });
        });
    });

    describe('mapDispatchToProps', () => {
        it('should map dispatch to props', () => {
            const dispatch = jest.fn();
            expect(JSON.stringify(mapDispatchToProps(dispatch))).toEqual(JSON.stringify({ volunteerActions: {} }));
        });
    });
    describe('render', () => {
        it('should render correctly', () => {
            expect(wrapper).toMatchSnapshot();
        });
    });
});
