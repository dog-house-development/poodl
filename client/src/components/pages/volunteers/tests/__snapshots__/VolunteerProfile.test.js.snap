// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VolunteerProfile tests render should render correctly 1`] = `
<div
  className="view-all-container"
>
  <Link
    className="button small tertiary"
    replace={false}
    to="/volunteers"
  >
    <i
      className="material-icons"
    >
      keyboard_backspace
    </i>
     Back to all volunteers
  </Link>
  <div>
    <h1>
      Big
       
      Tup
    </h1>
    <DynamicForm
      editValues={[Function]}
      editable={true}
      errors={Object {}}
      getValues={[Function]}
      inputs={
        Array [
          Object {
            "id": "Personal Info",
            "inputs": Array [
              Object {
                "id": "name",
                "inputs": Array [
                  Object {
                    "id": "firstName",
                    "kind": "field",
                    "label": "First name",
                    "style": Object {
                      "flexGrow": "5",
                    },
                    "type": "text",
                  },
                  Object {
                    "id": "middleInitial",
                    "kind": "field",
                    "label": "M.I.",
                    "style": Object {},
                  },
                  Object {
                    "id": "lastName",
                    "kind": "field",
                    "label": "Last name",
                    "style": Object {},
                    "type": "text",
                  },
                  Object {
                    "id": "nickName",
                    "kind": "field",
                    "label": "Nickname or Preferred Name",
                    "style": Object {},
                  },
                ],
                "kind": "flex",
              },
              Object {
                "id": "email",
                "kind": "field",
                "label": "Email",
                "type": "email",
              },
              Object {
                "editable": false,
                "id": "password",
                "kind": "field",
                "label": "Password",
                "type": "password",
              },
              Object {
                "editable": false,
                "id": "password2",
                "kind": "field",
                "label": "Confirm Password",
                "type": "password",
              },
              Object {
                "id": "address",
                "inputs": Array [
                  Object {
                    "id": "streetAddress",
                    "kind": "field",
                    "label": "Street Address",
                  },
                  Object {
                    "id": "city",
                    "kind": "field",
                    "label": "City",
                  },
                  Object {
                    "id": "state",
                    "kind": "field",
                    "label": "State",
                  },
                  Object {
                    "id": "zip",
                    "kind": "field",
                    "label": "ZIP",
                  },
                ],
                "kind": "flex",
              },
              Object {
                "id": "phones",
                "inputs": Array [
                  Object {
                    "id": "homePhone",
                    "kind": "field",
                    "label": "Home Phone",
                  },
                  Object {
                    "id": "cellPhone",
                    "kind": "field",
                    "label": "Cell or Message Phone",
                  },
                ],
                "kind": "flex",
              },
              Object {
                "id": "references",
                "kind": "field",
                "label": "References",
              },
            ],
            "kind": "group",
          },
          Object {
            "id": "Volunteer Info",
            "inputs": Array [
              Object {
                "id": "learnAboutVolunteerProgram",
                "kind": "radio",
                "label": "How did you learn about Belgrade Senior Center's volunteer program",
                "options": Array [
                  "Friend or Relative",
                  "Brochure/Poster",
                  "Newspaper, Television or Radio",
                  "School or College",
                  "Belgrade Senior Center",
                  "United Way",
                  "Community Event",
                  "Other",
                ],
              },
              Object {
                "id": "volunteerFrequency",
                "kind": "field",
                "label": "How often would you like to volunteer?",
              },
              Object {
                "id": "occasionalOneDayProjects",
                "kind": "selectBoolean",
                "label": "I would like to be contacted for occasional one-day projects that fit my skills and interests",
              },
              Object {
                "id": "singleDayActivity",
                "kind": "selectBoolean",
                "label": "I would like to volunteer for a single day activity.",
              },
              Object {
                "id": "availabilitySchedule",
                "kind": "field",
                "label": "What is your availability schedule?",
              },
            ],
            "kind": "group",
          },
          Object {
            "id": "skills",
            "inputs": Array [
              Object {
                "id": "houseMaintenanceAndRepairs",
                "kind": "multiCheckbox",
                "label": "Housing Maintenance and Repairs (weekday and weekend days)",
                "options": Array [
                  "Carpentry",
                  "Plumbing",
                  "Masonry",
                  "Cleaning",
                  "Electrical",
                  "Painting",
                ],
              },
              Object {
                "id": "groundMaintenance",
                "kind": "multiCheckbox",
                "label": "Grounds Maintenance (any time)",
                "options": Array [
                  "Lawn Maintenance",
                  "Grounds Cleanup",
                  "Pruning Trees and Shrubs",
                  "Planting and Maintaining Flower Beds",
                  "Snow Removal",
                ],
              },
              Object {
                "id": "clericalAssistance",
                "kind": "multiCheckbox",
                "label": "Clerical Assistance (week days)",
                "options": Array [
                  "Data Entry",
                  "Folding Brochures",
                  "General Office",
                  "Preparing Bulk Mailings",
                  "Front Desk",
                ],
              },
              Object {
                "id": "nutritionProgram",
                "kind": "multiCheckbox",
                "label": "Nutrition Program",
                "options": Array [
                  "Food Prep",
                  "Dishes",
                  "Deliver Meals on Wheels",
                  "Dining Room Setup",
                ],
              },
              Object {
                "id": "specialEventsAndFundRaising",
                "kind": "multiCheckbox",
                "label": "Special Events and Fundraising",
                "options": Array [
                  "Create Flyers, Brochures, and/or Posters",
                  "Assist with Events",
                  "Solicit Auction Items, Donations, Prizes, In-Kind Services",
                ],
              },
              Object {
                "id": "interpretingTeachingClassesOrWorkshops",
                "kind": "multiCheckbox",
                "label": "Interpreting, Teaching Classes or Workshops",
                "options": Array [
                  "Life Skill Classes",
                  "Painting, Crafts",
                  "Computer",
                  "Exercise",
                ],
              },
              Object {
                "id": "proposedActivities",
                "kind": "field",
                "label": "Please specify or propose any other volunteer opportunity",
              },
            ],
            "kind": "group",
            "label": "Skills and Interests (please check all that apply)",
          },
        ]
      }
      values={
        Object {
          "_id": "123",
          "email": "bigtup@nowhere.com",
          "firstName": "Big",
          "lastName": "Tup",
        }
      }
    />
  </div>
</div>
`;
